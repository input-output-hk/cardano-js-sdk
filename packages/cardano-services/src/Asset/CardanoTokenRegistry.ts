import { Asset, Cardano, ProviderError, ProviderFailure } from '@cardano-sdk/core';
import { InMemoryCache } from '../InMemoryCache';
import { Logger } from 'ts-log';
import { TokenMetadataService } from './types';
import axios, { AxiosInstance } from 'axios';

export const DEFAULT_TOKEN_METADATA_CACHE_TTL = 600;
export const DEFAULT_TOKEN_METADATA_SERVER_URL = 'https://tokens.cardano.org';

interface NumberValue {
  value?: number;
}

interface StringValue {
  value?: string;
}

interface TokenMetadataServiceRecord {
  decimals?: NumberValue;
  description?: StringValue;
  logo?: StringValue;
  name?: StringValue;
  subject?: string;
  ticker?: StringValue;
  url?: StringValue;
}

const propertiesToChange: Record<string, string> = { description: 'desc', logo: 'icon' };
export const toCoreTokenMetadata = (record: TokenMetadataServiceRecord) =>
  Object.fromEntries(
    Object.entries(record)
      .filter(([key]) => ['decimals', 'description', 'logo', 'name', 'ticker', 'url'].includes(key))
      .map(([key, value]) => [propertiesToChange[key] || key, value.value])
  ) as Asset.TokenMetadata;

const toProviderError = (error: unknown, details: string) => {
  if (error instanceof ProviderError) return error;

  const message = error instanceof Error ? `${error.message} ` : '';

  return new ProviderError(ProviderFailure.Unknown, error, `${message}${details}`);
};

/**
 * Configuration options for CardanoTokenRegistry
 */
export interface CardanoTokenRegistryConfiguration {
  /**
   * The cache TTL in seconds. Default: 10 minutes.
   */
  tokenMetadataCacheTTL?: number;

  /**
   * The Cardano Token Registry public API base URL. Default: https://tokens.cardano.org
   */
  tokenMetadataServerUrl?: string;
}

interface CardanoTokenRegistryConfigurationWithRequired extends CardanoTokenRegistryConfiguration {
  tokenMetadataCacheTTL: number;
  tokenMetadataServerUrl: string;
}

/**
 * Dependencies that are need to create CardanoTokenRegistry
 */
export interface CardanoTokenRegistryDependencies {
  /**
   * The cache engine. Default: InMemoryCache with CardanoTokenRegistryConfiguration.cacheTTL as default TTL
   */
  cache?: InMemoryCache;

  /**
   * The logger object
   */
  logger: Logger;
}

/**
 * TokenMetadataService implementation using Cardano Token Registry public API
 */
export class CardanoTokenRegistry implements TokenMetadataService {
  /**
   * The axios client used to retrieve metadata from API
   */
  #axiosClient: AxiosInstance;

  /**
   * The in memory cache engine
   */
  #cache: InMemoryCache;

  /**
   * The logger object
   */
  #logger: Logger;

  constructor({ cache, logger }: CardanoTokenRegistryDependencies, config: CardanoTokenRegistryConfiguration = {}) {
    const defaultConfig: CardanoTokenRegistryConfigurationWithRequired = {
      tokenMetadataCacheTTL: DEFAULT_TOKEN_METADATA_CACHE_TTL,
      tokenMetadataServerUrl: DEFAULT_TOKEN_METADATA_SERVER_URL,
      ...config
    };

    this.#cache = cache || new InMemoryCache(defaultConfig.tokenMetadataCacheTTL);
    this.#axiosClient = axios.create({ baseURL: defaultConfig.tokenMetadataServerUrl });
    this.#logger = logger;
  }

  shutdown() {
    this.#cache.shutdown();
  }

  async getTokenMetadata(assetIds: Cardano.AssetId[]) {
    this.#logger.debug(`Requested asset metadata for "${assetIds}"`);

    const [assetIdsToRequest, tokenMetadata] = this.getTokenMetadataFromCache(assetIds);

    // All metadata was taken from cache
    if (assetIdsToRequest.length === 0) return tokenMetadata;

    this.#logger.debug(`Fetching asset metadata for "${assetIdsToRequest}"`);

    try {
      const response = await this.#axiosClient.post<{ subjects: TokenMetadataServiceRecord[] }>('metadata/query', {
        properties: ['decimals', 'description', 'logo', 'name', 'ticker', 'url'],
        subjects: assetIdsToRequest
      });

      for (const record of response.data.subjects) {
        try {
          const { subject } = record;

          if (subject) {
            const assetId = Cardano.AssetId(subject);
            const metadata = toCoreTokenMetadata(record);

            tokenMetadata[assetIds.indexOf(assetId)] = metadata;
            this.#cache.set(assetId, metadata);
          } else
            throw new ProviderError(
              ProviderFailure.InvalidResponse,
              undefined,
              `Missing 'subject' property in metadata record ${JSON.stringify(record)}`
            );
        } catch (error) {
          throw toProviderError(error, `while evaluating metadata record ${JSON.stringify(record)}`);
        }
      }
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new ProviderError(
          ProviderFailure.ConnectionFailure,
          error,
          'CardanoTokenRegistry failed to fetch asset metadata from the token registry server'
        );
      }

      throw error;
    }

    return tokenMetadata;
  }

  getTokenMetadataFromCache(assetIds: Cardano.AssetId[]) {
    const assetIdsToRequest: Cardano.AssetId[] = [];
    // eslint-disable-next-line unicorn/no-new-array
    const cachedTokenMetadata: (Asset.TokenMetadata | null)[] = new Array(assetIds.length).fill(null);

    for (const [i, assetId] of assetIds.entries()) {
      const cachedMetadata = this.#cache.getVal<Asset.TokenMetadata>(assetId);

      if (cachedMetadata) {
        this.#logger.debug(`Using cached asset metadata value for "${assetId}"`);
        cachedTokenMetadata[i] = cachedMetadata;
      } else assetIdsToRequest.push(assetId);
    }

    return [assetIdsToRequest, cachedTokenMetadata] as const;
  }
}
