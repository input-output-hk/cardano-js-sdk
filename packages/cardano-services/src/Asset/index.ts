export * from './AssetHttpService';
export * from './CardanoTokenRegistry';
export * from './StubTokenMetadataService';
export * from './DbSyncNftMetadataService';
export * from './DbSyncAssetProvider';
export * from './TypeOrmNftMetadataService';
export * from './types';
