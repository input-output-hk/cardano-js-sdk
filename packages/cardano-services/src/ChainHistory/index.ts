export * from './ChainHistoryHttpService.js';
export * from './DbSyncChainHistory/index.js';
