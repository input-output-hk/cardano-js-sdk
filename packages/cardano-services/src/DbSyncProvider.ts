import { HealthCheckResponse, Provider } from '@cardano-sdk/core';
import { Pool } from 'pg';

const HEALTH_CHECK_QUERY = 'SELECT 1';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyArgs = any[];

export const DbSyncProvider = <
  T extends (abstract new (...args: AnyArgs) => {}) | (new (...args: AnyArgs) => {}) = { new (): {} }
>(
  BaseClass?: T
) => {
  abstract class Mixin extends (BaseClass || Object) implements Provider {
    public db: Pool;

    constructor(...args: AnyArgs) {
      const [db, ...baseArgs] = [...args];

      super(...baseArgs);

      this.db = db;
    }

    public async healthCheck(): Promise<HealthCheckResponse> {
      const result = await this.db.query(HEALTH_CHECK_QUERY);
      return { ok: !!result.rowCount };
    }
  }

  type BaseArgs = T extends abstract new (...baseArgs: infer A) => {}
    ? A
    : T extends new (...baseArgs: infer A) => {}
    ? A
    : never;
  type BaseInstance = T extends abstract new (...baseArgs: AnyArgs) => infer I
    ? I
    : T extends new (...baseArgs: AnyArgs) => infer I
    ? I
    : never;
  type ReturnedType = BaseInstance & { db: Pool; healthCheck: () => Promise<HealthCheckResponse> };

  return Mixin as unknown as (T extends new (...baseArgs: AnyArgs) => {}
    ? new (db: Pool, ...args: BaseArgs) => ReturnedType
    : abstract new (db: Pool, ...args: BaseArgs) => ReturnedType) & {
    prototype: { healthCheck: () => Promise<HealthCheckResponse> };
  };
};

export type DbSyncProvider<T> = T extends new (...args: AnyArgs) => infer O ? O : never;
