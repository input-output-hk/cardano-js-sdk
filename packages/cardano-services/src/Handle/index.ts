export * from './HandleHttpService';
export * from './TypeOrmHandleProvider';
