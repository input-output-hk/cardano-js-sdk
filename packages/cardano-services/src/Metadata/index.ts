export * from './types.js';
export * from './DbSyncMetadataService.js';
export * from './mappers.js';
export * as MetadataQueries from './queries.js';
