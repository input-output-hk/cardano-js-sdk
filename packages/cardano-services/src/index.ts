export * from './Asset';
export * from './Http';
export * from './InMemoryCache';
export * from './Program';
export * from './RunnableModule';
export * from './StakePool';
export * from './TxSubmit';
export * from './Utxo';
export * from './ChainHistory';
export * from './Rewards';
export * as errors from './errors';
export * as util from './util';
