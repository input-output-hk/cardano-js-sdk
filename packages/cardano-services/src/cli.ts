#!/usr/bin/env node
import {
  API_URL_DEFAULT,
  HttpServerOptions,
  OGMIOS_URL_DEFAULT,
  ProgramOptionDescriptions,
  RABBITMQ_URL_DEFAULT,
  SERVICE_DISCOVERY_BACKOFF_FACTOR_DEFAULT,
  SERVICE_DISCOVERY_TIMEOUT_DEFAULT,
  ServiceNames,
  loadHttpServer
} from './Program';
import { Command } from 'commander';
import {
  CommonOptionDescriptions,
  ENABLE_METRICS_DEFAULT,
  Programs,
  USE_QUEUE_DEFAULT,
  WrongOption
} from './ProgramsCommon';
import { DB_CACHE_TTL_DEFAULT } from './InMemoryCache';
import { DEFAULT_TOKEN_METADATA_CACHE_TTL, DEFAULT_TOKEN_METADATA_SERVER_URL } from './Asset';
import { EPOCH_POLL_INTERVAL_DEFAULT } from './NetworkInfo';
import { InvalidLoggerLevel } from './errors';
import {
  PARALLEL_MODE_DEFAULT,
  PARALLEL_TXS_DEFAULT,
  POLLING_CYCLE_DEFAULT,
  TxWorkerOptionDescriptions,
  TxWorkerOptions,
  loadAndStartTxWorker
} from './TxWorker';
import { URL } from 'url';
import { cacheTtlValidator } from './util/validators';
import { loggerMethodNames } from './util';
import clear from 'clear';
import fs from 'fs';
import onDeath from 'death';
import path from 'path';

const copiedPackageJsonPath = path.join(__dirname, 'original-package.json');
// Exists in dist/, doesn't exist when run with ts-node
const packageJsonPath = fs.existsSync(copiedPackageJsonPath)
  ? copiedPackageJsonPath
  : path.join(__dirname, '../package.json');
const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));

clear();
// eslint-disable-next-line no-console
console.log('Cardano Services CLI');

const stringToBoolean = (value: string, program: Programs, option: string) => {
  // for compatibility: accepting same values as envalid in startWorker.ts
  if (['0', 'f', 'false'].includes(value)) return false;
  if (['1', 't', 'true'].includes(value)) return true;
  throw new WrongOption(program, option, ['false', 'true']);
};

const commonOptions = (command: Command) =>
  command
    .option(
      '--logger-min-severity <level>',
      CommonOptionDescriptions.LoggerMinSeverity,
      (level) => {
        if (!loggerMethodNames.includes(level)) {
          throw new InvalidLoggerLevel(level);
        }
        return level;
      },
      'info'
    )
    .option(
      '--ogmios-url <ogmiosUrl>',
      CommonOptionDescriptions.OgmiosUrl,
      (url) => new URL(url),
      new URL(OGMIOS_URL_DEFAULT)
    )
    .option(
      '--rabbitmq-url <rabbitMQUrl>',
      CommonOptionDescriptions.RabbitMQUrl,
      (url) => new URL(url),
      new URL(RABBITMQ_URL_DEFAULT)
    )
    .option('--ogmios-srv-service-name <ogmiosSrvServiceName>', ProgramOptionDescriptions.OgmiosSrvServiceName)
    .option('--rabbitmq-srv-service-name <rabbitmqSrvServiceName>', ProgramOptionDescriptions.RabbitMQSrvServiceName)
    .option(
      '--service-discovery-backoff-factor <serviceDiscoveryBackoffFactor>',
      ProgramOptionDescriptions.ServiceDiscoveryBackoffFactor,
      (factor) => Number.parseFloat(factor),
      SERVICE_DISCOVERY_BACKOFF_FACTOR_DEFAULT
    )
    .option(
      '--service-discovery-timeout <serviceDiscoveryTimeout>',
      ProgramOptionDescriptions.ServiceDiscoveryTimeout,
      (interval) => Number.parseInt(interval, 10),
      SERVICE_DISCOVERY_TIMEOUT_DEFAULT
    );

const program = new Command('cardano-services');

program.version(packageJson.version);

commonOptions(
  program
    .command('start-server')
    .description('Start the HTTP server')
    .argument('<serviceNames...>', `List of services to attach: ${Object.values(ServiceNames).toString()}`)
)
  .option('--api-url <apiUrl>', ProgramOptionDescriptions.ApiUrl, (url) => new URL(url), new URL(API_URL_DEFAULT))
  .option('--enable-metrics', ProgramOptionDescriptions.EnableMetrics, () => true, ENABLE_METRICS_DEFAULT)
  .option(
    '--postgres-connection-string <postgresConnectionString>',
    ProgramOptionDescriptions.PostgresConnectionString,
    (url) => new URL(url).toString()
  )
  .option('--cardano-node-config-path <cardanoNodeConfigPath>', ProgramOptionDescriptions.CardanoNodeConfigPath)
  .option('--db-cache-ttl <dbCacheTtl>', ProgramOptionDescriptions.DbCacheTtl, cacheTtlValidator, DB_CACHE_TTL_DEFAULT)
  .option(
    '--epoch-poll-interval <epochPollInterval>',
    ProgramOptionDescriptions.EpochPollInterval,
    (interval) => Number.parseInt(interval, 10),
    EPOCH_POLL_INTERVAL_DEFAULT
  )
  .option(
    '--token-metadata-cache-ttl <tokenMetadataCacheTTL>',
    ProgramOptionDescriptions.TokenMetadataCacheTtl,
    cacheTtlValidator,
    DEFAULT_TOKEN_METADATA_CACHE_TTL
  )
  .option(
    '--token-metadata-server-url <tokenMetadataServerUrl>',
    ProgramOptionDescriptions.TokenMetadataServerUrl,
    (url) => new URL(url).toString(),
    DEFAULT_TOKEN_METADATA_SERVER_URL
  )
  .option('--use-queue', ProgramOptionDescriptions.UseQueue, () => true, USE_QUEUE_DEFAULT)
  .option('--postgres-srv-service-name <postgresSrvServiceName>', ProgramOptionDescriptions.PostgresSrvServiceName)
  .option('--postgres-db <postgresDb>', ProgramOptionDescriptions.PostgresDb)
  .option('--postgres-user <postgresUser>', ProgramOptionDescriptions.PostgresUser)
  .option('--postgres-password <postgresPassword>', ProgramOptionDescriptions.PostgresPassword)
  .option('--postgres-ssl-ca-file <postgresSslCaFile>', ProgramOptionDescriptions.PostgresSslCaFile)
  .action(async (serviceNames: ServiceNames[], options: { apiUrl: URL } & HttpServerOptions) => {
    const { apiUrl, ...rest } = options;
    try {
      const server = await loadHttpServer({ apiUrl: apiUrl || API_URL_DEFAULT, options: rest, serviceNames });
      await server.initialize();
      await server.start();
      onDeath(async () => {
        await server.shutdown();
        process.exit(1);
      });
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(error);
      process.exit(1);
    }
  });

commonOptions(program.command('start-worker').description('Start RabbitMQ worker'))
  .option(
    '--parallel [parallel]',
    TxWorkerOptionDescriptions.Parallel,
    (parallel) => stringToBoolean(parallel, Programs.RabbitmqWorker, TxWorkerOptionDescriptions.Parallel),
    PARALLEL_MODE_DEFAULT
  )
  .option(
    '--parallel-txs <parallelTxs>',
    TxWorkerOptionDescriptions.ParallelTxs,
    (parallelTxs) => Number.parseInt(parallelTxs, 10),
    PARALLEL_TXS_DEFAULT
  )
  .option(
    '--polling-cycle <pollingCycle>',
    TxWorkerOptionDescriptions.PollingCycle,
    (pollingCycle) => Number.parseInt(pollingCycle, 10),
    POLLING_CYCLE_DEFAULT
  )
  .action(async (options: TxWorkerOptions) => {
    // eslint-disable-next-line no-console
    console.log(`RabbitMQ transactions worker: ${options.parallel ? 'parallel' : 'serial'} mode`);
    const txWorker = await loadAndStartTxWorker({ options });
    onDeath(async () => {
      await txWorker.stop();
      process.exit(1);
    });
  });

if (process.argv.slice(2).length === 0) {
  program.outputHelp();
  process.exit(1);
} else {
  program.parseAsync(process.argv).catch((error) => {
    // eslint-disable-next-line no-console
    console.error(error);
    process.exit(1);
  });
}
