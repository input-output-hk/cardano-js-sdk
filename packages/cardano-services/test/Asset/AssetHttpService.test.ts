import {
  AssetHttpService,
  CardanoTokenRegistry,
  DbSyncAssetProvider,
  DbSyncNftMetadataService,
  HttpServer,
  HttpServerConfig,
  NftMetadataService,
  TokenMetadataService
} from '../../src';
import { AssetProvider, Cardano } from '@cardano-sdk/core';
import { CreateHttpProviderConfig, assetInfoHttpProvider } from '@cardano-sdk/cardano-services-client';
import { INFO, createLogger } from 'bunyan';
import { Pool } from 'pg';
import { createDbSyncMetadataService } from '../../src/Metadata';
import { getPort } from 'get-port-please';
import { dummyLogger as logger } from 'ts-log';
import { mockTokenRegistry } from './CardanoTokenRegistry.test';
import axios from 'axios';

const APPLICATION_JSON = 'application/json';
const APPLICATION_CBOR = 'application/cbor';
const UNSUPPORTED_MEDIA_STRING = 'Request failed with status code 415';
const BAD_REQUEST_STRING = 'Request failed with status code 400';

describe('AssetHttpService', () => {
  let apiUrlBase: string;
  let assetProvider: AssetProvider;
  let closeMock: () => Promise<void> = jest.fn();
  let config: HttpServerConfig;
  let db: Pool;
  let httpServer: HttpServer;
  let ntfMetadataService: NftMetadataService;
  let service: AssetHttpService;
  let port: number;
  let tokenMetadataServerUrl = '';
  let tokenMetadataService: TokenMetadataService;
  let clientConfig: CreateHttpProviderConfig<AssetProvider>;
  let provider: AssetProvider;

  beforeAll(async () => {
    port = await getPort();
    apiUrlBase = `http://localhost:${port}/asset`;
    config = { listen: { port } };
    clientConfig = { baseUrl: apiUrlBase, logger: createLogger({ level: INFO, name: 'unit tests' }) };
    provider = assetInfoHttpProvider(clientConfig);
  });

  describe('healthy state', () => {
    beforeAll(async () => {
      ({ closeMock, tokenMetadataServerUrl } = await mockTokenRegistry(() => ({})));
      db = new Pool({ connectionString: process.env.POSTGRES_CONNECTION_STRING });
      ntfMetadataService = new DbSyncNftMetadataService({
        db,
        logger,
        metadataService: createDbSyncMetadataService(db, logger)
      });
      tokenMetadataService = new CardanoTokenRegistry({ logger }, { tokenMetadataServerUrl });
      assetProvider = new DbSyncAssetProvider({ db, logger, ntfMetadataService, tokenMetadataService });
      service = new AssetHttpService({ assetProvider, logger });
      httpServer = new HttpServer(config, { services: [service] });
      await httpServer.initialize();
      await httpServer.start();
    });

    afterAll(async () => {
      await httpServer.shutdown();
      tokenMetadataService.shutdown();
      await db.end();
      await closeMock();
    });

    describe('/health', () => {
      it('/health response should be true', async () => {
        const res = await axios.post(`${apiUrlBase}/health`, undefined, {
          headers: { 'Content-Type': APPLICATION_JSON }
        });
        expect(res.status).toBe(200);
        expect(res.data).toEqual({ ok: true });
      });
    });

    describe('/get-asset', () => {
      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(`${apiUrlBase}/get-asset`, { args: [] }, { headers: { 'Content-Type': APPLICATION_CBOR } });
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });

      it('returns 400 coded response if the request is bad formed', async () => {
        try {
          await axios.post(`${apiUrlBase}/get-asset`, { args: [['test']] });
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.status).toBe(400);
          expect(error.message).toBe(BAD_REQUEST_STRING);
        }
      });

      it('returns 404 coded response for not existing existing asset id', async () => {
        try {
          const res = await axios.post(`${apiUrlBase}/get-asset`, {
            args: ['0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef']
          });
          expect(res.data[0]).toEqual({});
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } catch (error: any) {
          expect(error.response.status).toBe(404);
        }
      });

      it('returns asset info for existing asset id', async () => {
        const res = await provider.getAsset(
          Cardano.AssetId('50fdcdbfa3154db86a87e4b5697ae30d272e0bbcfa8122efd3e301cb6d616361726f6e2d63616b65')
        );
        expect(res).toMatchSnapshot();
      });

      it('returns asset info with extra data when requested', async () => {
        const res = await provider.getAsset(
          Cardano.AssetId('50fdcdbfa3154db86a87e4b5697ae30d272e0bbcfa8122efd3e301cb6d616361726f6e2d63616b65'),
          { history: true, nftMetadata: true, tokenMetadata: true }
        );
        const { history, nftMetadata, tokenMetadata } = res;

        expect(res).toMatchSnapshot();
        expect(history).toHaveLength(1);
        expect(nftMetadata).toBeDefined();
        expect(tokenMetadata).toBeDefined();
      });
    });
  });
});
