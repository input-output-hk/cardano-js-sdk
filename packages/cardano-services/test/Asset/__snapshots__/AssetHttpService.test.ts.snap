// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetHttpService healthy state /get-asset returns asset info for existing asset id 1`] = `
Object {
  "assetId": "50fdcdbfa3154db86a87e4b5697ae30d272e0bbcfa8122efd3e301cb6d616361726f6e2d63616b65",
  "fingerprint": "asset1f0azzptnr8dghzjh7egqvdjmt33e3lz5uy59th",
  "mintOrBurnCount": 1,
  "name": "6d616361726f6e2d63616b65",
  "policyId": "50fdcdbfa3154db86a87e4b5697ae30d272e0bbcfa8122efd3e301cb",
  "quantity": 1n,
}
`;

exports[`AssetHttpService healthy state /get-asset returns asset info with extra data when requested 1`] = `
Object {
  "assetId": "50fdcdbfa3154db86a87e4b5697ae30d272e0bbcfa8122efd3e301cb6d616361726f6e2d63616b65",
  "fingerprint": "asset1f0azzptnr8dghzjh7egqvdjmt33e3lz5uy59th",
  "history": Array [
    Object {
      "quantity": 1n,
      "transactionId": "f66791a0354c43d8c5a93671eb96d94633e3419f3ccbb0a00c00a152d3b6ca06",
    },
  ],
  "mintOrBurnCount": 1,
  "name": "6d616361726f6e2d63616b65",
  "nftMetadata": Object {
    "description": Array [
      "This is my first NFT of the macaron cake",
    ],
    "files": undefined,
    "image": Array [
      "ipfs://QmcDAmZubQig7tGUgEwbWcgdvz4Aoa2EiRZyFoX3fXTVmr",
    ],
    "mediaType": undefined,
    "name": "macaron cake token",
    "otherProperties": Map {
      "id" => 1n,
    },
    "version": "1.0",
  },
  "policyId": "50fdcdbfa3154db86a87e4b5697ae30d272e0bbcfa8122efd3e301cb",
  "quantity": 1n,
  "tokenMetadata": Object {
    "desc": "This is my first NFT of the macaron cake",
    "name": "macaron cake token",
  },
}
`;
